# Cloud Cost Anomaly Detection & Forecasting Using Data Analytics

* With enterprises rapidly scaling their operations on cloud platforms (AWS, Azure, GCP), cloud bills have become a major controllable cost factor. Many IT companies face:  
    * Unexpected spikes in usage leading to surprise bills.
    * Underutilized or idle resources consuming budget.
    * Anomalies caused by misconfigurations, attacks, or runaway jobs.
* Business leaders (CFOs, CTOs, Engineering Managers) need data-driven visibility into usage and spend patterns to forecast budgets accurately, detect anomalies early, and enforce governance.
* This project’s objective is to demonstrate how data analytics can reduce waste, improve efficiency, and prevent unexpected cloud overspend.

**Business Problem**
* Cloud cost overruns and anomalies lead to:
    * 20–40% overspending annually due to lack of governance and transparency.
    * Difficulty identifying which team/service/region caused the spike.
    * Inability to predict future spend accurately, leading to budget shocks.
    * SLA breaches when anomalies impact customer-facing workloads.
* Without robust anomaly detection and forecasting, IT companies lose millions and undermine operational efficiency.

**Goal**
* To build an end-to-end analytics solution that:
    * Detects and flags anomalous cloud spending in near real-time.
    * Forecasts future spend using statistical and machine learning models.
    * Identifies cost-saving opportunities by analyzing underutilized resources.
    * Provides actionable dashboards for both executives (strategic view) and engineers (operational root-cause analysis).


## Data Cleaning

## 1. Importing Libraries & Loading Dataset

In [1]:
# Since, the dataset consists of more than 1 million records, we would be loading the data in chunks
# Will convert it into parquet for faster re-reads and optimization

import pandas as pd

chunksize = 200_000
chunks = pd.read_csv('cloud_usage_data.csv', chunksize=chunksize)

df_list = []
for chunk in chunks:
    df_list.append(chunk)
df = pd.concat(df_list, ignore_index=True)
print(df.shape)

(1005000, 13)


In [2]:
# Converting to parquet for faster re-reads in later phases

df.to_parquet('cloud_usage_datafile.parquet', index=False)

In [3]:
# loading the full dataset at once

df = pd.read_parquet('cloud_usage_datafile.parquet')

In [4]:
df.head(10)

timestamp resource_id   service_tag service_canonical  \
0  2024-10-10T06:32:31.665035    9f546ae5         objs3                S3   
1  2025-06-08T02:19:47.665035    78e0807a  compute_east               EC2   
2  2025-07-26T04:59:10.665035    cf9713a0  compute_east               EC2   
3  2025-08-31T21:54:40.665035    71c42bb1           EC2               EC2   
4  2025-03-10T12:51:34.665035    5ddba1b8           EBS               EBS   
5  2025-09-19T16:15:57.665035    e1eb527d  loadbalancer               ELB   
6  2025-04-05T08:24:31.665035    a2ec574d            fn            Lambda   
7  2025-01-17T03:38:18.665035    16aa964c            fn            Lambda   
8  2025-08-03T22:09:36.665035    b609411d    db-managed               RDS   
9  2025-04-15T18:15:01.665035    d0a92372           cdn        CloudFront   

       region      team instance_type       usage  usage_unit  cost_usd  \
0   us-west-2    search     t3.medium    1.294475    GB-hours  0.000305   
1   us-west-2    search      m5.large    5.549555  vCPU-hours       NaN   
2   eu-west-1     infra      m5.large   11.416323  vCPU-hours  0.472819   
3   eu-west-1      None    spot-small    0.479407  vCPU-hours  0.028337   
4  ap-south-1      None      r5.large   21.487007    GB-hours  0.010587   
5   eu-west-1  payments    spot-small  124.000000    requests  0.115082   
6  ap-south-1  payments     c5.xlarge  383.042823    requests  0.000063   
7   eu-west-1    search      r5.large  191.000000    requests  0.000040   
8   us-east-1  devtools     c5.xlarge    0.535388  vCPU-hours  0.041171   
9   us-east-1  payments     t3.medium   21.101344    GB-hours  0.002464   

  currency              owner        source_ip  
0      INR             udavis    210.90.217.45  
1      INR           mrosario   22.219.141.112  
2      EUR           cjackson    208.132.8.249  
3      INR  harringtontimothy     49.94.27.144  
4      INR            jason04   122.212.132.65  
5      INR      walterherrera    27.143.221.40  
6      INR        connorhardy  182.213.230.111  
7      USD          charles65      205.9.71.87  
8      INR            james16    45.27.149.136  
9      INR     thompsondeanna    57.239.87.105

In [5]:
df.shape

(1005000, 13)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005000 entries, 0 to 1004999
Data columns (total 13 columns):
 #   Column             Non-Null Count    Dtype  
---  ------             --------------    -----  
 0   timestamp          1005000 non-null  object 
 1   resource_id        1005000 non-null  object 
 2   service_tag        1005000 non-null  object 
 3   service_canonical  1005000 non-null  object 
 4   region             1005000 non-null  object 
 5   team               781255 non-null   object 
 6   instance_type      1005000 non-null  object 
 7   usage              1005000 non-null  float64
 8   usage_unit         1005000 non-null  object 
 9   cost_usd           968604 non-null   float64
 10  currency           1005000 non-null  object 
 11  owner              1005000 non-null  object 
 12  source_ip          1005000 non-null  object 
dtypes: float64(2), object(11)
memory usage: 99.7+ MB


In [7]:
df.isna().sum()

timestamp                 0
resource_id               0
service_tag               0
service_canonical         0
region                    0
team                 223745
instance_type             0
usage                     0
usage_unit                0
cost_usd              36396
currency                  0
owner                     0
source_ip                 0
dtype: int64

In [8]:
df.sample(5)

timestamp resource_id   service_tag  \
683567  2025-03-21T20:33:07.665035    1b438a8e        lambda   
906978  2025-08-02T12:43:06.665035    76469127           rds   
977544  2025-08-06T19:01:32.665035    940504cb  loadbalancer   
60565   2025-02-08T14:37:38.665035    30c14207           cdn   
880488  2025-06-13T01:50:53.665035    6f2800a8           elb   

       service_canonical      region      team instance_type       usage  \
683567            Lambda   eu-west-1      None      r5.large   27.000000   
906978               RDS   us-east-1  payments    spot-small    3.476940   
977544               ELB  ap-south-1      None  custom-small  163.000000   
60565         CloudFront   us-east-1    search  custom-small   20.602359   
880488               ELB   us-east-1        ml     t3.medium   20.000000   

        usage_unit  cost_usd currency         owner        source_ip  
683567    requests  0.000005      INR    xrodriguez  219.249.208.134  
906978  vCPU-hours  0.251879      USD  jessicacross    163.37.47.187  
977544    requests  0.164699      INR      jackie07   56.215.246.128  
60565     GB-hours  0.002369      INR    smithharry     91.121.29.35  
880488    requests  0.020908      INR      chadford   157.65.170.235

In [9]:
# Converting timestamp into proper format

df['timestamp'] = pd.to_datetime(df['timestamp'], errors='coerce')

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005000 entries, 0 to 1004999
Data columns (total 13 columns):
 #   Column             Non-Null Count    Dtype         
---  ------             --------------    -----         
 0   timestamp          994932 non-null   datetime64[ns]
 1   resource_id        1005000 non-null  object        
 2   service_tag        1005000 non-null  object        
 3   service_canonical  1005000 non-null  object        
 4   region             1005000 non-null  object        
 5   team               781255 non-null   object        
 6   instance_type      1005000 non-null  object        
 7   usage              1005000 non-null  float64       
 8   usage_unit         1005000 non-null  object        
 9   cost_usd           968604 non-null   float64       
 10  currency           1005000 non-null  object        
 11  owner              1005000 non-null  object        
 12  source_ip          1005000 non-null  object        
dtypes: datetime64[ns](1), float

In [11]:
df.isna().sum()

timestamp             10068
resource_id               0
service_tag               0
service_canonical         0
region                    0
team                 223745
instance_type             0
usage                     0
usage_unit                0
cost_usd              36396
currency                  0
owner                     0
source_ip                 0
dtype: int64

In [12]:
# Checking for duplicates

df.duplicated().sum()

4955

In [13]:
duplicates_all = df[df.duplicated(keep=False)]
duplicates_all

timestamp resource_id   service_tag  \
107     2025-05-22 17:27:24.665035    ead3637d           k8s   
148     2025-02-26 23:22:39.665035    5e73d41f        Lambda   
233     2025-04-10 18:38:08.665035    c45b0c99           cdn   
326     2025-02-13 19:12:54.665035    c99395b7  compute_east   
361     2025-07-13 19:11:02.665035    a7694f84           eks   
...                            ...         ...           ...   
1004387 2025-03-31 17:55:04.665035    9b9bddb3           eks   
1004832 2025-07-24 01:47:50.665035    34adb899           cdn   
1004852                        NaT    9c99711b           eks   
1004934 2025-07-19 21:01:17.665035    5e42f96a           EKS   
1004949 2025-03-10 05:22:07.665035    25e70658           RDS   

        service_canonical      region            team instance_type  \
107                   EKS  ap-south-1            None      r5.large   
148                Lambda   eu-west-1        devtools     r6g.large   
233            CloudFront   us-west-2            None     r6g.large   
326                   EC2   eu-west-1        devtools    spot-small   
361                   EKS   us-east-1  recommendation     r6g.large   
...                   ...         ...             ...           ...   
1004387               EKS   us-west-2        devtools    spot-small   
1004832        CloudFront  ap-south-1          search      r5.large   
1004852               EKS   eu-west-1            None    spot-small   
1004934               EKS   us-east-1        payments  custom-small   
1004949               RDS   us-west-2        payments     c5.xlarge   

              usage  usage_unit  cost_usd currency            owner  \
107        0.575294  vCPU-hours  0.009738      INR         marywood   
148      537.675414    requests  0.000115      EUR         smithjon   
233        6.138191    GB-hours  0.000729      EUR          amber11   
326        0.416361  vCPU-hours  0.021255      USD   cervantesjames   
361        2.075104  vCPU-hours  0.044329      USD  christinewilson   
...             ...         ...       ...      ...              ...   
1004387    0.638073  vCPU-hours  0.013739      INR           agomez   
1004832    8.991063    GB-hours  0.000999      EUR      morsejeremy   
1004852    3.735201  vCPU-hours  0.062287      INR           jeff24   
1004934    3.310956  vCPU-hours  0.060383      INR   alvaradojustin   
1004949    7.581241  vCPU-hours  0.584938      EUR        bethany09   

               source_ip  
107      219.123.249.254  
148       218.149.251.36  
233         204.50.245.7  
326       164.205.127.66  
361        143.76.18.234  
...                  ...  
1004387     192.41.74.22  
1004832  193.219.147.123  
1004852   215.251.122.45  
1004934    204.148.44.69  
1004949      3.53.214.55  

[9910 rows x 13 columns]

In [14]:
df.groupby(df.columns.tolist()).size().reset_index(name='count').query('count > 1')

timestamp resource_id    service_tag  \
139    2024-09-24 11:57:52.665035    f6e64973            ELB   
304    2024-09-24 14:38:45.665035    0ead4c0c            rds   
504    2024-09-24 16:57:11.665035    99eff05f  block-storage   
508    2024-09-24 16:58:43.665035    5d668533             fn   
570    2024-09-24 17:37:51.665035    53943a92            cdn   
...                           ...         ...            ...   
740582 2025-09-23 20:34:35.665035    b942889b            EKS   
740615 2025-09-23 20:48:26.665035    9967291d            EBS   
740663 2025-09-23 21:04:19.665035    54d207cf            ec2   
741189 2025-09-24 01:28:44.665035    793a59b1     db-managed   
741690 2025-09-24 08:57:17.665035    3da6add0            EKS   

       service_canonical      region            team instance_type  \
139                  ELB   us-west-2       analytics      r5.large   
304                  RDS  ap-south-1          search  custom-small   
504                  EBS   us-east-1       analytics  custom-small   
508               Lambda   eu-west-1           infra      r5.large   
570           CloudFront  ap-south-1        devtools  custom-small   
...                  ...         ...             ...           ...   
740582               EKS  ap-south-1          search  custom-small   
740615               EBS  ap-south-1  recommendation      r5.large   
740663               EC2  ap-south-1  recommendation  custom-small   
741189               RDS   eu-west-1              ml    spot-small   
741690               EKS   us-west-2              ml     c5.xlarge   

             usage  usage_unit  cost_usd currency            owner  \
139     149.000000    requests  0.167503      EUR          ychavez   
304       0.713191  vCPU-hours  0.054245      USD  michaelwilliams   
504      15.042269    GB-hours  0.006155      USD     blackmatthew   
508      80.000000    requests  0.000014      USD  katherineosborn   
570       4.383106    GB-hours  0.000405      INR        ruizmaria   
...            ...         ...       ...      ...              ...   
740582   26.661430  vCPU-hours  0.639346      INR     floydheather   
740615   17.084278    GB-hours  0.006885      INR         nicole48   
740663    1.277596  vCPU-hours  0.054820      EUR     laurabonilla   
741189    2.765165  vCPU-hours  0.277479      USD         zsimpson   
741690    4.304157  vCPU-hours  0.080705      INR         aramirez   

              source_ip  count  
139       203.244.35.60      2  
304      39.152.104.119      2  
504        4.36.172.247      2  
508       96.173.231.36      2  
570        176.191.63.9      2  
...                 ...    ...  
740582    133.161.45.98      2  
740615  218.141.152.174      2  
740663   219.69.106.166      2  
741189       95.8.227.9      2  
741690   141.254.46.168      2  

[3658 rows x 14 columns]

In [15]:
df[df.duplicated(subset=['timestamp', 'resource_id', 'usage', 'cost_usd'], keep=False)]

timestamp resource_id   service_tag  \
107     2025-05-22 17:27:24.665035    ead3637d           k8s   
148     2025-02-26 23:22:39.665035    5e73d41f        Lambda   
233     2025-04-10 18:38:08.665035    c45b0c99           cdn   
326     2025-02-13 19:12:54.665035    c99395b7  compute_east   
361     2025-07-13 19:11:02.665035    a7694f84           eks   
...                            ...         ...           ...   
1004387 2025-03-31 17:55:04.665035    9b9bddb3           eks   
1004832 2025-07-24 01:47:50.665035    34adb899           cdn   
1004852                        NaT    9c99711b           eks   
1004934 2025-07-19 21:01:17.665035    5e42f96a           EKS   
1004949 2025-03-10 05:22:07.665035    25e70658           RDS   

        service_canonical      region            team instance_type  \
107                   EKS  ap-south-1            None      r5.large   
148                Lambda   eu-west-1        devtools     r6g.large   
233            CloudFront   us-west-2            None     r6g.large   
326                   EC2   eu-west-1        devtools    spot-small   
361                   EKS   us-east-1  recommendation     r6g.large   
...                   ...         ...             ...           ...   
1004387               EKS   us-west-2        devtools    spot-small   
1004832        CloudFront  ap-south-1          search      r5.large   
1004852               EKS   eu-west-1            None    spot-small   
1004934               EKS   us-east-1        payments  custom-small   
1004949               RDS   us-west-2        payments     c5.xlarge   

              usage  usage_unit  cost_usd currency            owner  \
107        0.575294  vCPU-hours  0.009738      INR         marywood   
148      537.675414    requests  0.000115      EUR         smithjon   
233        6.138191    GB-hours  0.000729      EUR          amber11   
326        0.416361  vCPU-hours  0.021255      USD   cervantesjames   
361        2.075104  vCPU-hours  0.044329      USD  christinewilson   
...             ...         ...       ...      ...              ...   
1004387    0.638073  vCPU-hours  0.013739      INR           agomez   
1004832    8.991063    GB-hours  0.000999      EUR      morsejeremy   
1004852    3.735201  vCPU-hours  0.062287      INR           jeff24   
1004934    3.310956  vCPU-hours  0.060383      INR   alvaradojustin   
1004949    7.581241  vCPU-hours  0.584938      EUR        bethany09   

               source_ip  
107      219.123.249.254  
148       218.149.251.36  
233         204.50.245.7  
326       164.205.127.66  
361        143.76.18.234  
...                  ...  
1004387     192.41.74.22  
1004832  193.219.147.123  
1004852   215.251.122.45  
1004934    204.148.44.69  
1004949      3.53.214.55  

[9910 rows x 13 columns]

In [16]:
df[df['resource_id'] == 'ead3637d']

timestamp resource_id service_tag service_canonical  \
107   2025-05-22 17:27:24.665035    ead3637d         k8s               EKS   
58793 2025-05-22 17:27:24.665035    ead3637d         k8s               EKS   

           region  team instance_type     usage  usage_unit  cost_usd  \
107    ap-south-1  None      r5.large  0.575294  vCPU-hours  0.009738   
58793  ap-south-1  None      r5.large  0.575294  vCPU-hours  0.009738   

      currency     owner        source_ip  
107        INR  marywood  219.123.249.254  
58793      INR  marywood  219.123.249.254

In [17]:
# Dropping duplicate records

before = df.shape[0]
df = df.drop_duplicates(subset=['timestamp', 'resource_id', 'usage', 'cost_usd'])
after = df.shape[0]
print('Removed Duplicates:', before - after)

Removed Duplicates: 4955


In [18]:
# Canonicalizing categorical fields

service_map = {
    'ec2':'EC2', 'compute-east':'EC2', 'compute_east':'EC2',
    'rds':'RDS', 'db-managed':'RDS',
    's3':'S3', 'object-store':'S3', 'objs3':'S3',
    'lambda':'Lambda', 'fn':'Lambda',
    'eks':'EKS', 'k8s':'EKS',
    'ebs':'EBS', 'block-storage':'EBS',
    'elb':'ELB', 'loadbalancer':'ELB',
    'cloudfront':'CloudFront', 'cdn':'CloudFront'
}

df['service_tag'] = df['service_tag'].str.strip().str.lower().map(service_map).fillna(df['service_tag'])

In [19]:
# Handling missing values

df['cost_usd'] = pd.to_numeric(df['cost_usd'], errors='coerce')

df['team'] = df['team'].fillna('Unknown')

In [20]:
df[['cost_usd', 'team']].isna().sum()

cost_usd    36235
team            0
dtype: int64

In [21]:
df[((df['currency'] == 'USD') & (df['cost_usd'].isna()))]

timestamp resource_id service_tag service_canonical  \
1748    2025-02-06 00:36:04.665035    dbf381b9         ELB               ELB   
2140    2024-09-29 03:12:09.665035    5e4bac3e         EC2               EC2   
2941    2025-05-15 14:42:43.665035    301468aa          S3                S3   
3198    2025-05-17 16:39:42.665035    e755b330      Lambda            Lambda   
4002    2024-11-10 23:52:24.665035    56994929         RDS               RDS   
...                            ...         ...         ...               ...   
1002423 2025-04-09 19:22:30.665035    32d5b72e         EBS               EBS   
1002635 2024-11-30 23:07:17.665035    ce8a762b      Lambda            Lambda   
1002857 2025-04-17 02:02:21.665035    57aa382e      Lambda            Lambda   
1003292 2025-03-17 22:16:29.665035    85154d61         RDS               RDS   
1003910 2024-09-30 19:56:38.665035    fe39b37b         ELB               ELB   

             region            team instance_type       usage  usage_unit  \
1748      us-east-1  recommendation     r6g.large  210.000000    requests   
2140     ap-south-1        payments  custom-small    1.549277  vCPU-hours   
2941      eu-west-1        devtools  custom-small   27.538003    GB-hours   
3198      us-east-1         Unknown     r6g.large  103.955838    requests   
4002      us-west-2        payments     c5.xlarge    1.373021  vCPU-hours   
...             ...             ...           ...         ...         ...   
1002423   us-east-1           infra    spot-small   51.912213    GB-hours   
1002635   us-east-1              ml  custom-small   35.104106    requests   
1002857   us-east-1         Unknown     t3.medium  190.000000    requests   
1003292   us-west-2        payments     t3.medium    0.270836  vCPU-hours   
1003910   us-east-1           infra     t3.medium  550.321729    requests   

         cost_usd currency             owner        source_ip  
1748          NaN      USD    gilmoremichael   154.105.25.222  
2140          NaN      USD       melindapope   110.67.189.242  
2941          NaN      USD           ularson     19.45.184.17  
3198          NaN      USD       lindseyjose  213.122.114.199  
4002          NaN      USD     maryfernandez   211.127.61.247  
...           ...      ...               ...              ...  
1002423       NaN      USD    williamjackson   192.130.137.79  
1002635       NaN      USD          ijohnson     164.81.165.3  
1002857       NaN      USD  morrisonjonathan  146.255.153.210  
1003292       NaN      USD      kennedysandy   63.180.181.106  
1003910       NaN      USD        jeremylara    135.128.54.27  

[1022 rows x 13 columns]

In [22]:
# There are critical missing values with currency as USD
# Instead of dropping those rows, we will impute missing costs using service base price
# Estimate cost = usage*base_unit_price

# Setting approximate base unit prices
base_price_map = {
    'EC2':0.05, 'RDS':0.08, 'S3':0.0002, 'Lambda':0.0000002,
    'EKS':0.02, 'EBS':0.0005, 'ELB':0.001, 'CloudFront':0.0001
}

def impute_cost(row):
    if pd.isna(row['cost_usd']) and row['currency']=='USD':
        return row['usage']*base_price_map.get(row['service_tag'], 0.01)
    else:
        return row['cost_usd']

df['cost_usd'] = df.apply(impute_cost, axis=1)

In [23]:
df['cost_imputed_flag'] = df['cost_usd'].isna()

In [24]:
df.head()

timestamp resource_id service_tag service_canonical  \
0 2024-10-10 06:32:31.665035    9f546ae5          S3                S3   
1 2025-06-08 02:19:47.665035    78e0807a         EC2               EC2   
2 2025-07-26 04:59:10.665035    cf9713a0         EC2               EC2   
3 2025-08-31 21:54:40.665035    71c42bb1         EC2               EC2   
4 2025-03-10 12:51:34.665035    5ddba1b8         EBS               EBS   

       region     team instance_type      usage  usage_unit  cost_usd  \
0   us-west-2   search     t3.medium   1.294475    GB-hours  0.000305   
1   us-west-2   search      m5.large   5.549555  vCPU-hours       NaN   
2   eu-west-1    infra      m5.large  11.416323  vCPU-hours  0.472819   
3   eu-west-1  Unknown    spot-small   0.479407  vCPU-hours  0.028337   
4  ap-south-1  Unknown      r5.large  21.487007    GB-hours  0.010587   

  currency              owner       source_ip  cost_imputed_flag  
0      INR             udavis   210.90.217.45              False  
1      INR           mrosario  22.219.141.112               True  
2      EUR           cjackson   208.132.8.249              False  
3      INR  harringtontimothy    49.94.27.144              False  
4      INR            jason04  122.212.132.65              False

In [25]:
df['cost_usd'].isna().sum()

35213

In [26]:
df[((df['currency'] == 'USD') & (df['cost_usd'].isna()))]

Empty DataFrame
Columns: [timestamp, resource_id, service_tag, service_canonical, region, team, instance_type, usage, usage_unit, cost_usd, currency, owner, source_ip, cost_imputed_flag]
Index: []

In [27]:
# Checking remaining null rows

null_rows = df[df['cost_usd'].isna()]
print('Remaining null rows:', null_rows.shape[0])
null_rows[['service_tag', 'usage', 'currency']].value_counts().head(20)

Remaining null rows: 35213


service_tag  usage  currency
Lambda       13.0   INR         18
             1.0    EUR         16
ELB          28.0   INR         14
             27.0   INR         13
Lambda       3.0    EUR         13
ELB          1.0    INR         12
Lambda       14.0   EUR         12
ELB          54.0   EUR         12
             5.0    INR         11
             110.0  INR         11
Lambda       24.0   INR         11
ELB          3.0    INR         11
Lambda       18.0   INR         11
ELB          32.0   INR         11
             34.0   EUR         11
Lambda       1.0    INR         10
             8.0    EUR         10
             32.0   EUR         10
             78.0   EUR         10
ELB          120.0  EUR         10
Name: count, dtype: int64

In [28]:
unmapped_services = df[df['cost_usd'].isna()]['service_tag'].unique()
print("Unmapped services:", unmapped_services)

Unmapped services: ['EC2' 'EBS' 'CloudFront' 'EKS' 'ELB' 'S3' 'RDS' 'Lambda']


In [29]:
df[((df['currency'] == 'INR') & (df['cost_usd'].isna()))]

timestamp resource_id service_tag service_canonical  \
1       2025-06-08 02:19:47.665035    78e0807a         EC2               EC2   
358     2024-10-17 23:44:12.665035    cd959b0a         EC2               EC2   
495     2025-09-13 05:13:39.665035    a164dad5  CloudFront        CloudFront   
511     2025-08-03 02:04:06.665035    a27f0c4d          S3                S3   
585     2025-02-20 23:59:47.665035    1a5a9456          S3                S3   
...                            ...         ...         ...               ...   
1004587 2025-08-24 19:53:33.665035    17538922          S3                S3   
1004707 2025-07-07 16:24:28.665035    f87e9f30         EC2               EC2   
1004900 2025-03-01 23:31:30.665035    4c5388f2         EKS               EKS   
1004919 2024-12-11 21:56:34.665035    a0c31dd0      Lambda            Lambda   
1004922 2024-11-30 20:13:54.665035    71191162         EC2               EC2   

             region            team instance_type       usage  usage_unit  \
1         us-west-2          search      m5.large    5.549555  vCPU-hours   
358       us-west-2       analytics     t3.medium    7.627691  vCPU-hours   
495      ap-south-1  recommendation  custom-small   44.434353    GB-hours   
511      ap-south-1              ml      m5.large    1.951698    GB-hours   
585       us-west-2         Unknown     t3.medium   78.829885    GB-hours   
...             ...             ...           ...         ...         ...   
1004587   us-west-2        devtools      r5.large    2.741106    GB-hours   
1004707   us-west-2              ml    spot-small    1.573427  vCPU-hours   
1004900   us-east-1        devtools     t3.medium    2.770770  vCPU-hours   
1004919   us-west-2  recommendation     t3.medium  616.000000    requests   
1004922   eu-west-1       analytics  custom-small   10.234724  vCPU-hours   

         cost_usd currency           owner        source_ip  cost_imputed_flag  
1             NaN      INR        mrosario   22.219.141.112               True  
358           NaN      INR    ambercoleman    125.147.33.12               True  
495           NaN      INR         zmccall   158.99.202.171               True  
511           NaN      INR          nmason   104.92.213.145               True  
585           NaN      INR         lbenton    162.18.72.186               True  
...           ...      ...             ...              ...                ...  
1004587       NaN      INR  johnsongregory   155.233.52.201               True  
1004707       NaN      INR     yanglindsay     35.58.201.68               True  
1004900       NaN      INR     elizabeth12    146.44.18.225               True  
1004919       NaN      INR          adam64  213.168.183.157               True  
1004922       NaN      INR  schmidtjasmine    138.91.72.131               True  

[17730 rows x 14 columns]

In [30]:
df[((df['currency'] == 'EUR') & (df['cost_usd'].isna()))]

timestamp resource_id service_tag service_canonical  \
44      2024-12-02 04:10:09.665035    4251bba9         EBS               EBS   
154     2025-06-16 03:48:48.665035    c75cfc10         EC2               EC2   
180     2025-07-17 00:55:21.665035    49fcc591  CloudFront        CloudFront   
254     2025-09-15 22:59:33.665035    0df2d064         EKS               EKS   
324     2025-05-01 20:26:37.665035    6aef799a         ELB               ELB   
...                            ...         ...         ...               ...   
1004601 2025-01-28 09:13:23.665035    0c15b669         EBS               EBS   
1004625 2025-03-22 09:44:23.665035    8f7795dd  CloudFront        CloudFront   
1004875 2025-04-12 18:47:59.665035    208e5755         RDS               RDS   
1004879 2025-05-19 02:23:03.665035    31a316ec  CloudFront        CloudFront   
1004989 2025-07-06 16:15:34.665035    9fc1a166      Lambda            Lambda   

             region            team instance_type       usage  usage_unit  \
44       ap-south-1          search     c5.xlarge   53.636247    GB-hours   
154      ap-south-1  recommendation     c5.xlarge    2.914629  vCPU-hours   
180      ap-south-1         Unknown     t3.medium   63.093385    GB-hours   
254       us-east-1        devtools     c5.xlarge    1.602607  vCPU-hours   
324       us-west-2              ml      r5.large   67.000000    requests   
...             ...             ...           ...         ...         ...   
1004601   us-west-2       analytics      m5.large    8.256377    GB-hours   
1004625  ap-south-1       analytics     r6g.large    2.838611    GB-hours   
1004875   eu-west-1          search    spot-small    4.032408  vCPU-hours   
1004879   us-east-1           infra      m5.large   25.160334    GB-hours   
1004989   eu-west-1           infra      m5.large  103.412671    requests   

         cost_usd currency        owner        source_ip  cost_imputed_flag  
44            NaN      EUR   veronica32   220.95.191.109               True  
154           NaN      EUR      bdaniel    207.198.231.5               True  
180           NaN      EUR      tammy67    124.224.68.68               True  
254           NaN      EUR  laurenmoore   207.31.182.219               True  
324           NaN      EUR     hector06  125.176.127.235               True  
...           ...      ...          ...              ...                ...  
1004601       NaN      EUR      bgarcia   168.208.178.29               True  
1004625       NaN      EUR    uschwartz   85.223.204.218               True  
1004875       NaN      EUR       gina03   115.224.231.92               True  
1004879       NaN      EUR       ohardy    213.56.225.90               True  
1004989       NaN      EUR     zschmitt    173.152.125.8               True  

[17483 rows x 14 columns]

In [31]:
# Handling currency inconsistencies
# Convert to USD using approximate exchange rates

exchange_rates = {'USD':1, 'INR':0.012, 'EUR':1.07}

df['cost_usd'] = df.apply(
    lambda x: x['cost_usd']*exchange_rates.get(x['currency'], 1) if pd.notnull(x['cost_usd']) else x['cost_usd'], axis=1
)

df['currency'] = 'USD'

In [32]:
df.isna().sum()

timestamp            10021
resource_id              0
service_tag              0
service_canonical        0
region                   0
team                     0
instance_type            0
usage                    0
usage_unit               0
cost_usd             35213
currency                 0
owner                    0
source_ip                0
cost_imputed_flag        0
dtype: int64

In [33]:
before = df.shape[0]
df = df.dropna(subset=['timestamp'])
after = df.shape[0]

print(f"Dropped {before - after} rows with invalid/missing timestamps")

Dropped 10021 rows with invalid/missing timestamps


In [34]:
# Base prices in USD
base_price_map = {
    'EC2':0.05, 'RDS':0.08, 'S3':0.0002, 'Lambda':0.0000002,
    'EKS':0.02, 'EBS':0.0005, 'ELB':0.001, 'CloudFront':0.0001
}

# Exchange rates (already defined earlier)
exchange_rates = {'USD':1, 'INR':0.012, 'EUR':1.07}

# Median usage per service for backup imputation
service_usage_median = df.groupby('service_tag')['usage'].median()

def impute_cost(row):
    if pd.isna(row['cost_usd']):
        usage_val = row['usage'] if pd.notna(row['usage']) else service_usage_median.get(row['service_tag'], 1)
        base_price = base_price_map.get(row['service_tag'], 0.01)
        exch_rate = exchange_rates.get(row['currency'], 1)
        return usage_val * base_price * exch_rate
    return row['cost_usd']

df['cost_usd'] = df.apply(impute_cost, axis=1)

In [35]:
print("Remaining nulls per column:\n", df.isna().sum())
print("Rows flagged as imputed:", df['cost_imputed_flag'].sum())

Remaining nulls per column:
 timestamp            0
resource_id          0
service_tag          0
service_canonical    0
region               0
team                 0
instance_type        0
usage                0
usage_unit           0
cost_usd             0
currency             0
owner                0
source_ip            0
cost_imputed_flag    0
dtype: int64
Rows flagged as imputed: 34836


In [36]:
# Fixing corrupted resource IDs

df['resource_id'] = df['resource_id'].str[:8]

In [37]:
# Finding large cost spikes which are our anomalies of interest

q1 = df['cost_usd'].quantile(0.25)
q3 = df['cost_usd'].quantile(0.75)
iqr = q3 - q1
upper = q3 + 1.5*iqr

df['is_outlier'] = df['cost_usd'] > upper
df.head(20)

timestamp resource_id service_tag service_canonical  \
0  2024-10-10 06:32:31.665035    9f546ae5          S3                S3   
1  2025-06-08 02:19:47.665035    78e0807a         EC2               EC2   
2  2025-07-26 04:59:10.665035    cf9713a0         EC2               EC2   
3  2025-08-31 21:54:40.665035    71c42bb1         EC2               EC2   
4  2025-03-10 12:51:34.665035    5ddba1b8         EBS               EBS   
5  2025-09-19 16:15:57.665035    e1eb527d         ELB               ELB   
6  2025-04-05 08:24:31.665035    a2ec574d      Lambda            Lambda   
7  2025-01-17 03:38:18.665035    16aa964c      Lambda            Lambda   
8  2025-08-03 22:09:36.665035    b609411d         RDS               RDS   
9  2025-04-15 18:15:01.665035    d0a92372  CloudFront        CloudFront   
10 2025-05-26 00:21:37.665035    166c7f30         RDS               RDS   
11 2024-10-29 03:46:10.665035    e04eb9aa      Lambda            Lambda   
12 2025-02-25 02:44:14.665035    8872b2b1      Lambda            Lambda   
13 2025-03-18 14:36:10.665035    91431717         EBS               EBS   
14 2024-11-06 16:35:39.665035    4d346927         RDS               RDS   
15 2025-02-07 21:47:43.665035    b7b3a580         ELB               ELB   
16 2025-08-05 17:39:00.665035    ef8a0761         EKS               EKS   
17 2025-01-14 00:03:54.665035    7ac4203e         ELB               ELB   
18 2024-11-22 19:46:09.665035    a72edc04      Lambda            Lambda   
19 2025-04-15 09:33:27.665035    b5e28d82      Lambda            Lambda   

        region            team instance_type       usage  usage_unit  \
0    us-west-2          search     t3.medium    1.294475    GB-hours   
1    us-west-2          search      m5.large    5.549555  vCPU-hours   
2    eu-west-1           infra      m5.large   11.416323  vCPU-hours   
3    eu-west-1         Unknown    spot-small    0.479407  vCPU-hours   
4   ap-south-1         Unknown      r5.large   21.487007    GB-hours   
5    eu-west-1        payments    spot-small  124.000000    requests   
6   ap-south-1        payments     c5.xlarge  383.042823    requests   
7    eu-west-1          search      r5.large  191.000000    requests   
8    us-east-1        devtools     c5.xlarge    0.535388  vCPU-hours   
9    us-east-1        payments     t3.medium   21.101344    GB-hours   
10   us-west-2           infra    spot-small   21.761301  vCPU-hours   
11   us-east-1          search     c5.xlarge  259.327950    requests   
12   eu-west-1         Unknown  custom-small   47.000000    requests   
13  ap-south-1          search      m5.large   13.988567    GB-hours   
14   eu-west-1  recommendation      r5.large    2.583421  vCPU-hours   
15   eu-west-1       analytics      r5.large  102.000000    requests   
16   eu-west-1         Unknown    spot-small    4.384806  vCPU-hours   
17   us-east-1              ml    spot-small  770.000000    requests   
18   eu-west-1              ml     r6g.large   93.000000    requests   
19   eu-west-1              ml     t3.medium    9.000000    requests   

        cost_usd currency              owner        source_ip  \
0   3.660000e-06      USD             udavis    210.90.217.45   
1   2.774778e-01      USD           mrosario   22.219.141.112   
2   5.059163e-01      USD           cjackson    208.132.8.249   
3   3.400440e-04      USD  harringtontimothy     49.94.27.144   
4   1.270440e-04      USD            jason04   122.212.132.65   
5   1.380984e-03      USD      walterherrera    27.143.221.40   
6   7.560000e-07      USD        connorhardy  182.213.230.111   
7   4.000000e-05      USD          charles65      205.9.71.87   
8   4.940520e-04      USD            james16    45.27.149.136   
9   2.956800e-05      USD     thompsondeanna    57.239.87.105   
10  1.968389e+00      USD        dustinbates   94.247.112.118   
11  5.243000e-05      USD     davidrodriguez     11.24.49.205   
12  8.000000e-06      USD             mark73     55.38.20.252   
13  7.184000e-03     

In [38]:
df['is_outlier'].value_counts()

is_outlier
False    829595
True     160429
Name: count, dtype: int64

In [39]:
df = df.astype({
    'resource_id':'string',
    'service_tag':'string',
    'region':'string',
    'team':'string',
    'instance_type':'string',
    'owner':'string',
    'source_ip':'string'
})

In [40]:
print('Rows:', df.shape[0])
print('Missing values per column:\n', df.isna().sum())
df.describe(include='all')

Rows: 990024
Missing values per column:
 timestamp            0
resource_id          0
service_tag          0
service_canonical    0
region               0
team                 0
instance_type        0
usage                0
usage_unit           0
cost_usd             0
currency             0
owner                0
source_ip            0
cost_imputed_flag    0
is_outlier           0
dtype: int64


timestamp resource_id service_tag  \
count                          990024      990024      990024   
unique                            NaN      989878           8   
top                               NaN    c276d87d         ELB   
freq                              NaN           2      124646   
mean    2025-03-25 19:31:43.348280832         NaN         NaN   
min        2024-09-24 09:07:48.665035         NaN         NaN   
25%     2024-12-24 12:55:01.915035136         NaN         NaN   
50%     2025-03-25 19:00:40.665035008         NaN         NaN   
75%     2025-06-25 02:07:13.915035136         NaN         NaN   
max        2025-09-24 09:06:20.665035         NaN         NaN   
std                               NaN         NaN         NaN   

       service_canonical     region     team instance_type          usage  \
count             990024     990024   990024        990024  990024.000000   
unique                 8          4        8             7            NaN   
top                  ELB  us-west-2  Unknown      r5.large            NaN   
freq              124646     247965   220404        141783            NaN   
mean                 NaN        NaN      NaN           NaN      79.092907   
min                  NaN        NaN      NaN           NaN       0.007001   
25%                  NaN        NaN      NaN           NaN       3.804575   
50%                  NaN        NaN      NaN           NaN      13.452728   
75%                  NaN        NaN      NaN           NaN      55.483196   
max                  NaN        NaN      NaN           NaN  697303.713424   
std                  NaN        NaN      NaN           NaN    1529.195139   

       usage_unit       cost_usd currency   owner       source_ip  \
count      990024  990024.000000   990024  990024          990024   
unique          3            NaN        1  341895          989787   
top      GB-hours            NaN      USD  asmith  212.226.72.116   
freq       370698            NaN   990024     237               2   
mean          NaN       0.106787      NaN     NaN             NaN   
min           NaN       0.000000      NaN     NaN             NaN   
25%           NaN       0.000200      NaN     NaN             NaN   
50%           NaN       0.004176      NaN     NaN             NaN   
75%           NaN       0.063216      NaN     NaN             NaN   
max           NaN     675.786806      NaN     NaN             NaN   
std           NaN       1.901622      NaN     NaN             NaN   

       cost_imputed_flag is_outlier  
count             990024     990024  
unique                 2          2  
top                False      False  
freq              955188     829595  
mean                 NaN        NaN  
min                  NaN        NaN  
25%                  NaN        NaN  
50%                  NaN        NaN  
75%                  NaN        NaN  
max                  NaN        NaN  
std                  NaN        NaN

In [41]:
# Saving the cleaned file for EDA

df.to_parquet('clean_cloud_data.parquet', index=False)

### Conclusion

* In this phase, I worked on preparing a large scale synthetic dataset of approximately 1 million records of cloud usage and cost logs for analysis.
* The dataset was messy, containing missing values, duplicate records, inconsistent service names, corrupted timestamps, and outlier cost values.
* Through systematic data cleaning, I was able to achieve the following:
    * Standardized timestamps: Removed ~10,000 invalid entries.
    * Removed duplicates: Ensured data integrity across the data
    * Canonicalized categorical fields: mapped inconsistent service names into standard categories (EC2, RDS, S3, etc.)
    * Handled missing values: Filled missing team information with 'Unknown' and applied advanced imputation on approximately 35,000 missing cost entries using a combination of base service prices, usage values, and currency exchange rates.
    * Standardized currencies: Converted INR and EUR values into USD.
    * Cleaned resource IDs and owner fields: corrected corrupted identifiers.
    * Flagged anomalies: Identified 16% of rows as cost outliers for further investigation.
* After cleaning, the dataset now contains 989,961 rows with no missing values across any column. It is standardized, consistent, and fully ready for exploratory data analysis and anomaly detection.
* In this phase, I was able to demonstrate the importance of robust data preparation in real-world IT analytics projects, where messy, inconsistent, and incomplete datasets are the norm.
* The cleaned dataset provides a reliable foundation to extract meaningful insights in the next phase of the project.